File: Classes/EVEOnlineAPI/Character/EVECharMedals.m
//
//  EVECharMedals.m
//  EVEOnlineAPI
//
//  Created by Artem Shimanski on 6/4/10.
//  Copyright 2010 __MyCompanyName__. All rights reserved.
//

#import "EVECharMedals.h"

@implementation EVECharMedal

+ (NSDictionary*) scheme {
	static NSDictionary* scheme = nil;
	if (!scheme)
		scheme = @{@"medalID":@{@"type":@(EVEXMLSchemePropertyTypeScalar)},
				   @"reason":@{@"type":@(EVEXMLSchemePropertyTypeString)},
				   @"status":@{@"type":@(EVEXMLSchemePropertyTypeString)},
				   @"issuerID":@{@"type":@(EVEXMLSchemePropertyTypeScalar)},
				   @"issued":@{@"type":@(EVEXMLSchemePropertyTypeDate)}};
	
	return scheme;
}


@end


@implementation EVECharOtherCorporationsMedal

+ (NSDictionary*) scheme {
	static NSDictionary* scheme = nil;
	if (!scheme) {
		scheme = @{@"corporationID":@{@"type":@(EVEXMLSchemePropertyTypeScalar)},
				   @"title":@{@"type":@(EVEXMLSchemePropertyTypeString)},
				   @"medalDescription":@{@"type":@(EVEXMLSchemePropertyTypeString), @"elementName":@"description"}};
		
		NSMutableDictionary* dic = [[super scheme] mutableCopy];
		[dic setValuesForKeysWithDictionary:scheme];
		scheme = dic;
	}
	
	return scheme;
}

@end


@implementation EVECharMedals

+ (NSDictionary*) scheme {
	static NSDictionary* scheme = nil;
	if (!scheme)
		scheme = @{@"currentCorporation":@{@"type":@(EVEXMLSchemePropertyTypeRowset), @"class":[EVECharMedal class]},
				   @"otherCorporations":@{@"type":@(EVEXMLSchemePropertyTypeRowset), @"class":[EVECharOtherCorporationsMedal class]}};
	
	return scheme;
}

@end

File: EVEAPI/EVEAPI/EVECharMedals.swift
//
//  EVECharMedals.swift
//  EVEAPI
//
//  Created by Artem Shimanski on 29.11.16.
//  Copyright © 2016 Artem Shimanski. All rights reserved.
//

import UIKit

public class EVECharMedalsItem: EVEObject {
	public var medalID: Int = 0
	public var reason: String = ""
	public var status: EVEMedalStatus = .Public
	public var issuerID: Int = 0
	public var issued: Date = Date.distantPast
	public var corporationID: Int = 0
	public var title: String = ""
	public var medalDescription: String = ""
	
	public required init?(dictionary:[String:Any]) {
		super.init(dictionary: dictionary)
	}
	
	public required init?(coder aDecoder: NSCoder) {
		super.init(coder: aDecoder)
	}
	
	override public func scheme() -> [String:EVESchemeElementType] {
		return [
			"medalID":EVESchemeElementType.Int(elementName:nil, transformer:nil),
			"reason":EVESchemeElementType.String(elementName:nil, transformer:nil),
			"status":EVESchemeElementType.Int(elementName:nil, transformer:{(value:Any?) -> Any? in
				if let s = value as? String {
					return EVEMedalStatus(s).rawValue
				}
				else {
					return EVEMedalStatus.Public.rawValue
				}
			}),
			"issuerID":EVESchemeElementType.Int(elementName:nil, transformer:nil),
			"issued":EVESchemeElementType.Date(elementName:nil, transformer:nil),
			"corporationID":EVESchemeElementType.Int(elementName:nil, transformer:nil),
			"title":EVESchemeElementType.String(elementName:nil, transformer:nil),
			"medalDescription":EVESchemeElementType.String(elementName:"description", transformer:nil),
		]
	}
}


public class EVECharMedals: EVEResult {
	public var currentCorporation: [EVECharMedalsItem] = []
	public var otherCorporations: [EVECharMedalsItem] = []
	
	public required init?(dictionary:[String:Any]) {
		super.init(dictionary: dictionary)
	}
	
	public required init?(coder aDecoder: NSCoder) {
		super.init(coder: aDecoder)
	}
	
	override public func scheme() -> [String:EVESchemeElementType] {
		return [
			"currentCorporation":EVESchemeElementType.Rowset(elementName: nil, type: EVECharMedalsItem.self, transformer: nil),
			"otherCorporations":EVESchemeElementType.Rowset(elementName: nil, type: EVECharMedalsItem.self, transformer: nil),
		]
	}
}
